src/sink.rs

```diff
@@ -0,0 +1,847 @@
+//! Provides a `Sink` interface for publishing to hedwig.
+
+use crate::{Message, PublishBatch, Publisher, Topic, ValidatedMessage};
+use either::Either;
+use futures_util::{ready, sink::Sink, stream::Stream};
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// A sink which ingests messages, validates them with the given validator, then publishes them
+/// with the given publisher.
+///
+/// This is a convenience function which combines [`validator_sink`](validator_sink) and
+/// [`publisher_sink`](publisher_sink)
+pub fn sink<M, P>(
+    validator: M::Validator,
+    publisher: P,
+    batch_size: usize,
+) -> ValidatorSink<M, M::Validator, PublisherSink<P, P::PublishStream>>
+where
+    M: Message,
+    P: Publisher,
+    P::PublishStream: Unpin,
+{
+    // TODO a retry layer could go between these two
+    validator_sink(validator, publisher_sink(publisher, batch_size))
+}
+
+/// A sink which ingests messages, validates them with the given validator, then forwards them to
+/// the given destination sink.
+pub fn validator_sink<M, S>(
+    validator: M::Validator,
+    destination_sink: S,
+) -> ValidatorSink<M, M::Validator, S>
+where
+    M: Message,
+    S: Sink<(Topic, ValidatedMessage)>,
+{
+    ValidatorSink {
+        _message_type: std::marker::PhantomData,
+        validator,
+        sink: destination_sink,
+    }
+}
+
+/// A sink which ingests validated messages and publishes them to the given publisher.
+///
+/// This sink internally batches elements to publish multiple messages at once. The `batch_size`
+/// argument can be adjusted to control the number of elements stored in these batches.
```

Comment:
I think that the "wait for a full batch" is ultimately a worse choice for a batching scheme in almost all situations. In particular:

1. It has non-trivial latency implications (up-to infinity): if I sink 9 messages, but my batch is configured as 10, I would rather it to still just work somehow, rather than see everything hang;
2. It has non-trivial throughput implications: publishers may segment the batches as necessary to satisfy the API quotas and constraints. It seems fairly easy to set up the batch size in a way that would force very unfortunate segmentations to happen. On the other end of the spectrum, message batch may contain multiple messages destined for different topics – low batch sizes can lead to publisher putting very low number of messages in each API request, resulting in reduced throughput.

Reply:
I'll expand on the description a bit. To summarize, it waits for a full batch before sending, but that's only with `poll_ready`. Users (or stacked sink layers) can request a flush themselves using `poll_flush`. That choice could happen for example by choosing to use [`feed`](https://docs.rs/futures/0.3.13/futures/sink/trait.SinkExt.html#method.feed) vs [`send`](https://docs.rs/futures/0.3.13/futures/sink/trait.SinkExt.html#method.send), or a `sink.send_all(stream.by_ref().take(N))`, etc. One could even write a Sink adapter that does the "partial buffer just works" you suggested, maybe using some timeout to flush elements.

To your broader concern, you're right that there are implications in latency and throughput. Ultimately those are difficult to balance in either direction. A user knows their workload better than the library, if they send 7 items at a time, then a library-chosen throughput-favoring value of 10 could be harmful. On the other hand, the user doesn't know what is best for the publisher's internal implementation.

_Maybe_ one improvement would be to have each publisher independently provide a sink implementation that batches in some optimal segment size. But that starts to get pretty complex and implementation driven, I'm not sure it would be worth it

Reply:
its quite common to have a time config param where elements are also popped/flushed if the batch is not full but the elements have been sitting there for longer than a certain period of time.

Reply:
a generic timeout flusher Sink is totally compatible with this as it exists (standardized interfaces are good!). Post-MVP work for sure though